### ISSUE TITLE
Cilium does not remove taint on already running nodes

### ISSUE THREAD

**Issue by rustrial** [cilium/cilium#17103]
## Bug report

Cilium does not remove the node taint `node.cilium.io/agent-not-ready` if the taint is set on existing node (after cilium agent started). We observe this on our AWS EKS clusters. After upgrading to Cilium `1.10.3` we added the taint to all EKS NodeGroups and EKS immediately applied the taint not only to newly started nodes but also to all already existing nodes. 

---

### Expected behaviour

Cilium should remove the taint (if cilium is ready) even if the taint is set (later) after startup. Cilium should be resilient and insistently remove the taint whenever it is added again.

### Observed behaviour 

Cilium only removes the taint if it was already set on startup. If the taint is added later to an already running node, it is not removed unless the cilium agent on that node is restarted. 

### Workaround

Manually restarting the cilium agent (` kubectl rollout restart -n kube-system daemonset/cilium`) resolved the problem. 

However, that workaround is cumbersome as it needs manual intervention, which breaks our GitOps flow and can lead to severe downtime as all nodes are not eligible for scheduling of workloads. 

---

**General Information**

- Cilium version `1.10.3`
- Kernel version `5.4.117-58.216.amzn2.x86_64`
- Orchestration system version in use `v1.19.13-eks-8df270`

**How to reproduce the issue**

1. Upgrade to or install Cilium 1.10.3 on AWS EKS cluster
2. Add node taint `node.cilium.io/agent-not-ready` to EKS NodeGroup
3. The taint is not removed until the cilium agent is restarted

**Comment by aanm** [comment]
@rustrial Thank you for opening the issue, can you provide a cilium sysdump of the broken cluster? The implemented behavior in Cilium should be to remove the taint if it is re-added after Cilium has started.

**Comment by rustrial** [comment]
@aanm I set `debug=true` and restarted the cilium agent at about 20:00:00, at around 20:01 I manually added the taint and it has not been remove yet (as of 20:05:00). [Here](https://gist.githubusercontent.com/rustrial/792ee58dab23525e40b28191c6237e85/raw/31f23f44e20345cd149c2ee50d4acf6b522dd94f/gistfile1.txt) you can view the log dump from the agent on the affected node for the time between 20:00 and 20:03.

**Comment by aanm** [comment]
@rustrial thanks for opening the GH issue, the bug will be fixed by #17112

**Comment by elmazzun** [comment]
This is happening again for me, after installing Cilium `v1.16.5` via Helm chart in a KinD cluster: my Nodes have `node.cilium.io/agent-not-ready` Taint.
I found it out when enabling Cilium mesh:

```bash
$ cilium-kind-lab git:(main) ✗ kubectl -n kube-system describe pod clustermesh-apiserver-5998f87c84-5plq8

Warning  FailedScheduling  66s (x2 over 6m13s)  default-scheduler
0/3 nodes are available:
1 node(s) had untolerated taint {node-role.kubernetes.io/control-plane: }
2 node(s) had untolerated taint {node.cilium.io/agent-not-ready: }.
preemption: 0/3 nodes are available:
3 Preemption is not helpful for scheduling.
```
